sound/soc/sof/intel/hda-dai.c

```diff
@@ -400,6 +400,101 @@ static const struct snd_soc_dai_ops hda_link_dai_ops = {
 	.prepare = hda_link_pcm_prepare,
 };
 
+static struct sof_ipc_dai_config* ssp_dai_config_control(struct snd_sof_dev *sdev,
+							 struct snd_sof_dai *dai)
+{
+	struct sof_ipc_dai_config *config;
+	struct snd_sof_control *scontrol;
+	struct sof_ipc_ctrl_data *cdata;
+	int number_configs = dai->number_configs;
+	int current_config = dai->current_config;
+	int index;
+
+	if (current_config > number_configs - 1) {
+		dev_err(sdev->dev, "error: dai config index out of range\n");
+		return NULL;
+	}
+
+	config = &dai->dai_config[current_config];
+
+	/* return the only possible config if only 1 config present */
+	if (number_configs < 2)
+		return config;
+
+	/* find out if there's a control associated with the dai */
+	list_for_each_entry(scontrol, &sdev->kcontrol_list, list) {
+		cdata = scontrol->control_data;
+		if (scontrol->comp_id == dai->comp_dai.comp.id) {
+			/* find out if this control is meant for choosing dai config */
+			if (cdata->cmd == SOF_CTRL_CMD_ENUM) {
+				index = cdata->chanv[0].value;
+				/* check that the control value is in range and for ssp only */
+				if (index < number_configs &&
+				    dai->dai_config[index].type == SOF_DAI_INTEL_SSP) {
+					dev_dbg(sdev->dev,
+						"found dai config at index %d\n", index);
+					config = &dai->dai_config[index];
+					dai->current_config = index;
+				}
+			}
+			break;
+		}
+	}
+
+	return config;
+}
+
+static int ssp_dai_hw_params(struct snd_pcm_substream *substream,
+			     struct snd_pcm_hw_params *params,
+			     struct snd_soc_dai *dai)
+{
+
+	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
+	struct snd_soc_component *component =
+		snd_soc_rtdcom_lookup(rtd, SOF_AUDIO_PCM_DRV_NAME);
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
+	struct sof_ipc_dai_config *config;
+	struct snd_sof_dai *sof_dai;
+	struct sof_ipc_reply reply;
+	int ret;
+
+	list_for_each_entry(sof_dai, &sdev->dai_list, list) {
+		if (!sof_dai->cpu_dai_name)
+			continue;
+
+		if (!strcmp(dai->name, sof_dai->cpu_dai_name) &&
+		    substream->stream == sof_dai->comp_dai.direction) {
+			config = ssp_dai_config_control(sdev, sof_dai);
```

Comment:
ok, I am lost here.

This is just selecting a config without looking at the hw_params.

What happens if the user sets the frame sync to 48kHz and the hw_params is 8Hz rate? 

I just don't get how errors might be handled.

Reply:
If you set your SSP to 48kHz and try to play 8khz to pcm ->  you get "input/output error" etc. so aplay doesn't work as the hw_params don't match the dai. Exactly similar functionality as you would get currently if you try to play without this selection mechanism to 48khz pipeline without SRC. If you have SRC in your pipe things just work.

Reply:
I thought we're going to match with runtime hw_params and match config that matches no?

Reply:
there will be 3 "part" mechanism: 1. check if there's related control 2. if no control, match to pcm params, 3 if no match, take default. And I'll move the match to fe fixup...